jumanji/environments/routing/connector/utils.py

```diff
@@ -273,3 +245,68 @@ def is_repeated_later(positions: chex.Array) -> chex.Array:
     # Step 3: Combine masks: True if arr[i] == arr[j] (as pairs) AND j > i then
     # check if any such preceding identical pair j exists.
     return jnp.any(is_equal_pair & is_next, axis=1)
+
+
+def get_surrounded_mask(grid: chex.Array) -> chex.Array:
+    """
+    Args:
+        grid: 2D JAX array where 0 = open cell, positive integers = occupied
+
+    Returns:
+        Boolean mask where True indicates a cell is surrounded by occupied cells
+    """
+
+    # Create binary occupancy map
+    occupied = (grid > 0).astype(jnp.float32)
+
+    # Kernel to check all 4 neighbors
+    kernel = jnp.array([[0, 1, 0], [1, 0, 1], [0, 1, 0]])
+
+    # Count occupied neighbors
+    neighbor_count = convolve2d(occupied, kernel, mode="same")
+
+    # Cell is surrounded if all possible neighbors are occupied
+    possible_max_neighbors = convolve2d(jnp.ones_like(occupied), kernel, mode="same")
+    surrounded = neighbor_count == possible_max_neighbors
+
+    return surrounded
+
+
+@functools.partial(jax.jit, static_argnums=(0,))
+def get_adjacency_mask(grid_shape: tuple, coordinate: jax.Array) -> jax.Array:
+    """
+    Creates a mask with 1s on cells adjacent to a coordinate.
+
+    Args:
+        grid_shape: A tuple (rows, cols) defining the grid dimensions.
+        coordinate: A JAX array or tuple (row, col) for the center point.
+
+    Returns:
+        A JAX array of shape `grid_shape` with 1s for adjacent cells
+        and 0s elsewhere.
+    """
+    # 1. Start with a grid of zeros
+    mask = jnp.zeros(grid_shape, dtype=jnp.int32)
```

Comment:
Seems to be a boolean in the end?
```suggestion
    mask = jnp.zeros(grid_shape, dtype=jnp.bool_)
```